Support configuring partition type UUIDs for GPT partition auto-discovery

berrange opened the pull request:
The discoverable partitions specification defines a way to identify partitions used for various well known mount points:

  https://uapi-group.org/specifications/specs/discoverable_partitions_specification/

This relies on setting well known GPT partition GUIDs, instead of parted's generic "Linux Data"  GUID which more or less everything gets set to today (swap being a notable exception).

My goal is to make it possible for applications like Anaconda to "do the right thing(tm)" with little to no effort when doing disk partitioning. This series aims to make that possible, while still retaining the current behaviour by default for the sake of backwards compatibility.

With this series, a simple app usage might not need anything more than to opt-in to discoverable partitions with:

```
from blivet.git import gpt_set_discoverable_partitions

gpt_set_discoverable_partitions(True)
```

Assuming the disk has 'gpt' disk label and the app is already passing 'mountpoint' to the ``blivet.new_partition()`` method then it will "just work" and pick the right GUID for the given mount point. The `mountpoint` parameter would historically only be needed if a format is requested with `new_partition`, but now it can be passed regardless so it can be forwarded onto the `PartitionDevice` constructor.

The `examples/partitioning.py` script illustrates this, as does a comprehensive test at ``tests/vmtests/gpt_test.py``

If the app is more advanced and creating `PartitionDevice` instances directly instead of via the convenient `blivet.py` API entrypoints, then it will need to be changed to pass `mountpoint` to this object constructor too, but this is still hopefully fairly straightforward. 

Note that this functionality relies on `pypartd` exposing the `type_uuid` attribute on the Partition object. This is new functionality added to pyparted last month in https://github.com/dcantrell/pyparted/pull/95  Note `pyparted` only exposes this if compiled against a new enough `parted`. As a result some code additions in this series have 'hasattr' changes to attempt to gracefully degrade when the functionality is missing.

Since this is my first time looking at Blivet code in any detail, its taken me a while to get my head around the design and expected usage patterns. I'm still not entirely confident in my understanding, but the `examples/partitioning.py` script is at least doing the right thing with this series of changes.

Happy to have feedback on things I've overlooked / approached in the wrong/suboptimal way.

StorageGhoul commented:
Can one of the admins verify this patch?

dwlehman commented:
[Disclaimer: I'm not an active developer here anymore, so I may miss important things.]

Looks like a really nice addition to me. Your comprehension of the bits you're using here seems fine. The only thing I'm unsure of is whether it's necessary to set/manage `PartitionDevice.req_part_type_uuid` in addition to `PartitionDevice._part_type_uuid`; @vojtechtrefny will be able to say definitively. Generally, the `req_` attributes are for 'saving' data that will be stored in parted objects, which may get modified or replaced in the course of partition allocation; I think since you store `_part_type_uuid` outside of the parted object you can skip its `req_` counterpart.

Other than that, it might be good to extend the attribute checking to the places where you're setting it in `blivet/formats/disklabel.py` and in `blivet/deviceaction.py`

Thanks for your contribution!

vojtechtrefny commented:
Jenkins, ok to test.

vojtechtrefny commented on blivet/gpt.py:
```diff
@@ -0,0 +1,228 @@
+        "powerpc": "ppc",
+    }
+
+    if arch is None:
```
You can use the `get_arch` function from the `blivet.arch` module here. (Probably more complicated than the `platform` module, but it's probably better to use something we already have.

berrange replied:
I'll have a look and see if it has the right semantics

vojtechtrefny commented on blivet/gpt.py:
```diff
@@ -0,0 +1,228 @@
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
```
The better place for this file would probably be `blivet/devicelibs`. Strictly speaking GPT is not a device, but this where we already have some definitions for LUKS or btrfs too.

berrange replied:
Sure, works for me.

vojtechtrefny commented on blivet/gpt.py:
```diff
@@ -0,0 +1,228 @@
+        "s390": uuid.UUID("17440e4f-a8d0-467f-a46e-3912ae6ef2c5"),
+        "s390x": uuid.UUID("3f324816-667b-46ae-86ee-9b0c0c6c11b4"),
+        "tilegx": uuid.UUID("4ede75e2-6ccc-4cc8-b9c7-70334b087510"),
+        "x86_64": uuid.UUID("e7bb33fb-06cf-4e81-8273-e543b413e2e2"),
```
This is more a note to myself: we should move these definitions to libblockdev in the future (possibly using the GUID list provided by libfdisk). We had similar discussion recently about these definitions in UDisks https://github.com/storaged-project/udisks/pull/1032 and it makes sense to have these in one place if possible. (But that's out of scope for this PR.)

berrange replied:
Yeah, it would make sense to have them in a place that can maximise sharing.

vojtechtrefny commented on blivet/gpt.py:
```diff
@@ -183,6 +183,18 @@
 }
 
 
+_gpt_discoverable_partitions = False
```
Please use `blivet.flags.Flags` here instead. We already have multiple flags for controlling various default behaviours, so it make more sense having this there too. Anaconda is also setting these flags, so it will make it easier for the installer too.

berrange replied:
Ah yes, that's a much better place for this.

vojtechtrefny reviewed (changes requested):
Looks good to me in general, I have just few nitpicks. Thank you for the contribution, it is a great improvement, this feature has been requested for years.

I agree with Dave about the `req_part_type_uuid`, which is probably not necessary (but we can keep it, if you think it's better or clearer that way). I also plan to do some manual testing, but everything looks to be covered by tests so I don't expect any problems.

dwlehman commented:
> WRT tracking req_part_type_uuid directly, I was having some trouble distinguishing the user specified UUIDs from the default UUIDs that parted sets. Essentially every parted partition reports a UUID all the time. What's even more complex is that while we set a custom UUID when creating the partition, when we later format the partition with a filesystem, parted will replace the UUID with a default one again. I'll have another look at this all though, to see if there's a saner way.

We aren't using parted to create file systems, so I don't think parted would have an opportunity to change the UUID at this point. It might be a bug in parted that it's doing that when you use it to create the file system.

berrange commented:
> We aren't using parted to create file systems, so I don't think parted would have an opportunity to change the UUID at this point. It might be a bug in parted that it's doing that when you use it to create the file system.

I presume you're referring to Anaconda in this context ?  Blivet at least has functionality for formatting filesystems using parted, which is where I was hitting the problems in testing. I've not actually tried anything with anaconda yet.

dwlehman commented:
> > We aren't using parted to create file systems, so I don't think parted would have an opportunity to change the UUID at this point. It might be a bug in parted that it's doing that when you use it to create the file system.
> 
> I presume you're referring to Anaconda in this context ? Blivet at least has functionality for formatting filesystems using parted, which is where I was hitting the problems in testing. I've not actually tried anything with anaconda yet.

Wow, I forgot all about that. it's in `ActionCreateFormat.execute`, just above where you are setting the UUID. We aren't using parted to create the file system, but we are telling it to take note of the file system type, which I forgot about. Maybe the best way is to expand the conditional that sets `self.device.parted_partition.system` so that it handles both cases in one place.

berrange commented:
> Wow, I forgot all about that. it's in `ActionCreateFormat.execute`, just above where you are setting the UUID. We aren't using parted to create the file system, but we are telling it to take note of the file system type, which I forgot about. Maybe the best way is to expand the conditional that sets `self.device.parted_partition.system` so that it handles both cases in one place.

The `ActionCreateFormat.execute` code is affecting UUID in several ways. The `self.device.parted_partition.system` property will change it, as you mention, but also calling `self.device.unset_flag` or `self.device.set_flag` methods will also change the UUID depending on which flag is touched, as some of the flags have associated UUID mappings defined in parted's `flag_uuid_mapping` table (see https://git.savannah.gnu.org/cgit/parted.git/tree/libparted/labels/gpt.c#n179).

I didn't feel it was sensible to try to second guess which flags might have side effects on the partition type UUID. So took the approach of just explicitly setting the partition type UUID at the end after everything else has been done, which should be reliable and clearer to understand.

berrange commented:
I've pushed an updated series

 * switched to blivet.get_arch  and use blivet arch names in GPT UUID lookup tables
 * Use blivet.flags for the global toggle for use of discoverable partitions
 * moved the gpt.py module under blivet.devicelibs
 * Made more use of hasattr() and raise exceptions if parted doesn't support setting the type UUID

berrange commented:
Also, I switched the base to be the 3.7-devel branch

berrange commented:
The Fedora 36 jenkins job failure looks like it might be an infra problem, if someone can re-trigger it

vojtechtrefny reviewed (approved):
Thank you, looks good to me.

vojtechtrefny commented:
> > Wow, I forgot all about that. it's in `ActionCreateFormat.execute`, just above where you are setting the UUID. We aren't using parted to create the file system, but we are telling it to take note of the file system type, which I forgot about. Maybe the best way is to expand the conditional that sets `self.device.parted_partition.system` so that it handles both cases in one place.
> 
> The `ActionCreateFormat.execute` code is affecting UUID in several ways. The `self.device.parted_partition.system` property will change it, as you mention, but also calling `self.device.unset_flag` or `self.device.set_flag` methods will also change the UUID depending on which flag is touched, as some of the flags have associated UUID mappings defined in parted's `flag_uuid_mapping` table (see https://git.savannah.gnu.org/cgit/parted.git/tree/libparted/labels/gpt.c#n179).
> 
> I didn't feel it was sensible to try to second guess which flags might have side effects on the partition type UUID. So took the approach of just explicitly setting the partition type UUID at the end after everything else has been done, which should be reliable and clearer to understand.

I think your solution to set the GUID after the flags is fine. The parted flags that are sometimes flags, sometimes GPT attributes, sometimes partition GUIDs or IDs were always confusing so setting the UUID after the flags should be safe.

vojtechtrefny commented:
> Also, I switched the base to be the 3.7-devel branch

Thanks. I have few more changes planned for the 3.7 release, but if everything goes as planned, it should be released in time for Fedora 38 beta.
